dbs-virtio-devices: fix Balloon::write_config() bugs

00xc opened the pull request:
Fix two bugs in `Balloon::write_config()`:
1. `offset + data_len` might overflow. This causes a panic in debug builds, and wraps around in release builds.
https://github.com/openanolis/dragonball-sandbox/blob/8ed5e3fea96bf16747945d65c2b386542b28e1b7/crates/dbs-virtio-devices/src/balloon.rs#L634
2. Source (`data`) and destination (`right`) slices' sizes might be mismatched in `right.copy_from_slice(data)`, which causes a panic ([this is documented](https://doc.rust-lang.org/std/primitive.slice.html#panics-29)). This can happen if `data` is shorter than the right side of `config_slice.split_at_mut(offset)`:
https://github.com/openanolis/dragonball-sandbox/blob/8ed5e3fea96bf16747945d65c2b386542b28e1b7/crates/dbs-virtio-devices/src/balloon.rs#L638-L640

00xc committed:
dbs-virtio-devices: fix Balloon::write_config() bugs

1. Fix potential integer overflow when calculating the end of the
write destination (`offset + data_len`).

2. Fix a potential panic when copying into the destination slice due
to size mismatch.

Signed-off-by: Carlos López <carlos.lopez@suse.com>

00xc commented:
See also firecracker-microvm/firecracker#3869

ZizhengBian commented:
Please fix ut/clippy

studychao commented:
@00xc Hi, could you please help fix the UT before this week since we are going to migrate all dbs crates to kata containers next week.

00xc commented:
Should be good now

studychao commented:
@00xc There seems several clippy warnings please help fix it.

Tip: we are using Rust 1.70.

00xc committed:
fix clippy warnings

Signed-off-by: Carlos López <carlos.lopez@suse.com>

00xc commented:
Those warnings popped up on code completely unrelated to this PR. I have fixed them, but that seems like a very unusual request.

00xc committed:
dbs-legacy-devices: make CmosDevice public

The new device is public but behind a private module. Add a `pub use`
for the exports of the module.

Signed-off-by: Carlos López <carlos.lopez@suse.com>

studychao commented:
@00xc , sorry, after dig into the reason for CI complaining for the code not in your PR, we found out that there is one PR merged without clippy pass. Do you mind fix that in this PR or I could raise a new PR to fix this ?

studychao commented:
@00xc Hi, the PR itself looks good to me. I'll handle the clippy warnings.

00xc commented:
> @00xc Hi, the PR itself looks good to me. I'll handle the clippy warnings.

Thanks!
